<launch>
  <!-- Device mode -->
  <arg name="device_color"      default="0" />
  <arg name="device_depth"      default="1" />
  <arg name="device_all"        default="2" />
  <!-- Color mode -->
  <arg name="color_raw"         default="0" />
  <arg name="color_rectified"   default="1" />
  <!-- Depth mode -->
  <arg name="depth_raw"         default="0" />
  <arg name="depth_gray"        default="1" />
  <arg name="depth_colorful"    default="2" />
  <!-- Stream mode -->
  <arg name="stream_640x480"    default="0" />
  <arg name="stream_1280x480"   default="1" />
  <arg name="stream_1280x720"   default="2" />
  <arg name="stream_2560x720"   default="3" />
  <!-- 8059 VGA-->
  <arg name="stream_640x400"    default="4" />
  <arg name="stream_custom"    default="5" />
  <!-- Stream format -->
  <arg name="stream_mjpg"       default="0" />
  <arg name="stream_yuyv"       default="1" />

  <!-- 0 --> MONO16 -->
  <!-- 1 --> TYPE_16UC1 -->
  <!-- 2 --> TYPE_RGB -->
  <arg name="type_mono16"        default="0" />
  <arg name="type_16uc1"         default="1" />
  <arg name="type_RGB"           default="2" />

  <!-- Depth data type -->
  <arg name="type_default"              default="0" />
  <arg name="type_8_bits"               default="1" />
  <arg name="type_14_bits"              default="2" />
  <arg name="type_8_bits_x80"           default="3" />
  <arg name="type_11_bits"              default="4" />
  <arg name="type_off_rectify"          default="5" />
  <arg name="type_8_bits_raw"           default="6" />
  <arg name="type_14_bits_raw"          default="7" />
  <arg name="type_8_bits_x80_raw"       default="8" />
  <arg name="type_11_bits_raw"          default="9" />
  <arg name="type_14_bits_comb_rectify" default="11" />
  <arg name="type_11_bits_comb_rectify" default="13" />

  <!-- IMU timestamp align -->
  <arg name="imu_timestamp_align" default="true" />

  <!-- Camera Params -->

  <!-- Device index -->
  <arg name="dev_index" default="0" />
  <!-- Framerate -->
  <arg name="framerate" default="30" />
  <arg name="low_framerate" default="0" />

  <!--
  Device mode
    device_color: left_color ✓ right_color ? depth x
    device_depth: left_color x right_color x depth ✓
    device_all:   left_color ✓ right_color ? depth ✓
  Note: ✓: available, x: unavailable, ?: depends on #stream_mode
  -->
  <arg name="dev_mode" default="$(arg device_all)" />

  <arg name="color_mode" default="$(arg color_raw)" />
  <!-- Note: must set DEPTH_RAW to get raw depth values for points -->
  <arg name="depth_mode" default="$(arg depth_raw)" />
  <arg name="stream_mode" default="$(arg stream_2560x720)" />
  <arg name="color_resolution" default="1280x720" />
  <arg name="depth_resolution" default="640x360" />
  <arg name="color_stream_format" default="$(arg stream_yuyv)" />
  <arg name="depth_stream_format" default="$(arg stream_yuyv)" />
  <arg name="depth_type" default="$(arg type_RGB)" />
  <arg name="depth_data_type" default="$(arg type_11_bits)" />
  <arg name="depth_maximum_mm" default="1000" />

  <!-- Auto-exposure -->
  <arg name="state_ae" default="true" />
  <!-- Auto-white balance -->
  <arg name="state_awb" default="true" />
  <!-- IR intensity -->
  <arg name="ir_intensity" default="3" />
  <!-- IR Depth Only -->
  <arg name="ir_depth_only" default="false" />

  <!-- Generating points frequency, make slower than framerate -->
  <arg name="points_frequency" default="0" />
  <!-- Points display z distance scale factor -->
  <arg name="points_factor" default="1000.0" />

  <!-- Setup your local gravity here -->
  <arg name="gravity" default="9.8" />

  <!-- Node params -->

  <arg name="dm_preview"       default="dm_preview" />
  <arg name="base_frame"    default="$(arg dm_preview)_link_frame" />

  <arg name="left_mono_frame"   default="$(arg dm_preview)_left_mono_frame" />
  <arg name="left_color_frame"  default="$(arg dm_preview)_left_color_frame" />
  <arg name="right_mono_frame"  default="$(arg dm_preview)_right_mono_frame" />
  <arg name="right_color_frame" default="$(arg dm_preview)_right_color_frame" />
  <arg name="depth_frame"   default="$(arg dm_preview)_depth_frame" />
  <arg name="points_frame"  default="$(arg dm_preview)_points_frame" />
  <arg name="imu_frame"     default="$(arg dm_preview)_imu_frame" />
  <arg name="temp_frame"    default="$(arg dm_preview)_temp_frame" />
  <arg name="imu_frame_processed"     default="$(arg dm_preview)_imu_frame_processed" />

  <!-- left topics -->
  <arg name="left_mono_topic"  default="$(arg dm_preview)/left/image_mono" />
  <arg name="left_color_topic" default="$(arg dm_preview)/left/image_color" />
  <!-- right topics -->
  <arg name="right_mono_topic"  default="$(arg dm_preview)/right/image_mono" />
  <arg name="right_color_topic" default="$(arg dm_preview)/right/image_color" />
  <!-- depth topic -->
  <arg name="depth_topic"   default="$(arg dm_preview)/depth/image_raw" />
  <!-- points topic -->
  <arg name="points_topic"  default="$(arg dm_preview)/points/data_raw" />
  <!-- imu topic origin -->
  <arg name="imu_topic"     default="$(arg dm_preview)/imu/data_raw" />
  <!-- temp topic -->
  <arg name="temp_topic"    default="$(arg dm_preview)/temp/data_raw" />
  <!-- imu topic processed -->
  <arg name="imu_processed_topic"     default="$(arg dm_preview)/imu/data_raw_processed" />

  <arg name="pi"   value="3.1415926535897932" />
  <arg name="pi/2" value="1.5707963267948966" />
  <arg name="optical_rotate" value="0 0 0 -$(arg pi/2) 0 -$(arg pi/2)" />

  <node pkg="tf" type="static_transform_publisher" name="left_mono_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg left_mono_frame) 100" />
  <node pkg="tf" type="static_transform_publisher" name="left_color_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg left_color_frame) 100" />

  <node pkg="tf" type="static_transform_publisher" name="right_mono_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg right_mono_frame) 100" />
  <node pkg="tf" type="static_transform_publisher" name="right_color_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg right_color_frame) 100" />

  <node pkg="tf" type="static_transform_publisher" name="depth_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg depth_frame) 100" />
  <node pkg="tf" type="static_transform_publisher" name="point_cloud_broadcaster"
      args="$(arg optical_rotate) $(arg base_frame) $(arg points_frame) 100" />
  <node pkg="tf" type="static_transform_publisher" name="imu_broadcaster"
      args="0 0 0 $(arg pi/2) 0 $(arg pi) $(arg base_frame) $(arg imu_frame) 50" />
  <node pkg="tf" type="static_transform_publisher" name="temp_broadcaster"
      args="0 0 0 $(arg pi/2) 0 $(arg pi) $(arg base_frame) $(arg temp_frame) 50" />
  <node pkg="tf" type="static_transform_publisher" name="imu_processed_broadcaster"
      args="0 0 0 $(arg pi/2) 0 $(arg pi) $(arg base_frame) $(arg imu_frame_processed) 50" />
  <arg name="mesh_file" default="Arrow.obj" />

  <node name="dm_preview_node" pkg="dm_preview" type="dm_preview_node" output="screen" cwd="node">

    <!-- Camera params -->
    <param name="dev_index"     value="$(arg dev_index)" />
    <param name="framerate"     value="$(arg framerate)" />
    <param name="low_framerate"    value="$(arg low_framerate)" />
    <param name="dev_mode"      value="$(arg dev_mode)" />

    <param name="color_resolution"      value="$(arg color_resolution)" />
    <param name="depth_resolution"      value="$(arg depth_resolution)" />

    <param name="color_mode"    value="$(arg color_mode)" />
    <param name="depth_mode"    value="$(arg depth_mode)" />
    <param name="stream_mode"   value="$(arg stream_mode)" />
    <param name="color_stream_format" value="$(arg color_stream_format)" />
    <param name="depth_stream_format" value="$(arg depth_stream_format)" />
    <param name="state_ae"      value="$(arg state_ae)" />
    <param name="state_awb"     value="$(arg state_awb)" />
    <param name="ir_intensity"  value="$(arg ir_intensity)" />
    <param name="ir_depth_only" value="$(arg ir_depth_only)" />
    <param name="imu_timestamp_align" value="$(arg imu_timestamp_align)" />

    <param name="points_factor"    value="$(arg points_factor)" />
    <param name="points_frequency" value="$(arg points_frequency)" />

    <param name="gravity" value="$(arg gravity)" />

    <param name="depth_type" value="$(arg depth_type)" />
    <param name="depth_data_type" value="$(arg depth_data_type)" />
    <param name="depth_maximum_mm"    value="$(arg depth_maximum_mm)" />

    <!-- Frame ids -->
    <param name="base_frame"   value="$(arg base_frame)" />
    <param name="left_mono_frame"   value="$(arg left_mono_frame)" />
    <param name="left_color_frame"  value="$(arg left_color_frame)" />
    <param name="right_mono_frame"  value="$(arg right_mono_frame)" />
    <param name="right_color_frame" value="$(arg right_color_frame)" />
    <param name="depth_frame"  value="$(arg depth_frame)" />
    <param name="points_frame" value="$(arg points_frame)" />
    <param name="imu_frame"    value="$(arg imu_frame)" />
    <param name="temp_frame"   value="$(arg temp_frame)" />
    <param name="imu_frame_processed"    value="$(arg imu_frame_processed)" />

    <!-- Topic names -->

    <param name="left_mono_topic"   value="$(arg left_mono_topic)" />
    <param name="left_color_topic"  value="$(arg left_color_topic)" />
    <param name="right_mono_topic"  value="$(arg right_mono_topic)" />
    <param name="right_color_topic" value="$(arg right_color_topic)" />
    <param name="depth_topic"       value="$(arg depth_topic)" />
    <param name="points_topic"      value="$(arg points_topic)" />
    <param name="imu_topic"         value="$(arg imu_topic)" />
    <param name="temp_topic"        value="$(arg temp_topic)" />
    <param name="imu_processed_topic"         value="$(arg imu_processed_topic)" />

    <param name="mesh_file" value="$(arg mesh_file)" />
  </node>

  <!-- disable compressed depth plugin for image topics -->
  <group ns="$(arg left_mono_topic)">
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressedDepth'
    </rosparam>
  </group>
  <group ns="$(arg left_color_topic)">
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressedDepth'
    </rosparam>
  </group>
  <group ns="$(arg right_mono_topic)">
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressedDepth'
    </rosparam>
  </group>
  <group ns="$(arg right_color_topic)">
    <rosparam param="disable_pub_plugins">
      - 'image_transport/compressedDepth'
    </rosparam>
  </group>
</launch>
